### PR TITLE
Optimizing NonMatching::create_coupling_mass_matrix

### PR DESCRIPTION
**Pull request by GivAlz**
Hi everyone,
 following changes the algorithm in the style of https://github.com/dealii/dealii/pull/7563 .

If anyone has suggestions (even on how to avoid using these cumbersome containers with vectors of vectors of vectors) I would be glad to hear it.

The other reason I post is the following: due to how the function works, I slightly changed the of an addition, thus a test in debug mode fails (differences are in the order 1e-14/-16): would it be ok to make it so the tests prints something like the order of the error?

@luca-heltai  this is my other PR, now I shall work on the distributed version of both!

Thank you,
Giovanni

Edit: I just did the changes I wanted to make and force pushed!

### CODE REVIEW

File: source/non_matching/coupling.cc [hunk 1]
```diff
@@ -346,10 +425,17 @@ namespace NonMatching
                               {
                                 // Get the corresponding q point
                                 const unsigned int q = ids[oq];
-
+                                //                              std::cout<< " Q
+                                //                              i j " << q << "
+                                //                              " << i << " " <<
+                                //                              j << std::endl;
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: Surely, you wanted to remove these comments, don't you?

File: source/non_matching/coupling.cc [hunk 2]
```diff
@@ -346,10 +425,17 @@ namespace NonMatching
...
                                 cell_matrix(i, j) +=
                                   (fe_v.shape_value(j, q) *
                                    o_fe_v.shape_value(i, oq) * fe_v.JxW(q));
+
+                                //                              std::cout<< "
+                                //                              Added " <<
+                                //                              std::endl;
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: Surely, you wanted to remove these comments, don't you?

File: source/non_matching/coupling.cc [hunk 3]
```diff
@@ -361,7 +447,9 @@ namespace NonMatching
                                                        matrix);
               }
           }
+        std::cout << "Finished with this cell" << std::endl;
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: Surely, you wanted to remove these comments, don't you?

File: source/non_matching/coupling.cc [hunk 4]
```diff
@@ -361,7 +447,9 @@ namespace NonMatching
...
       }
+    std::cout << "Function completed" << std::endl;
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: Surely, you wanted to remove these comments, don't you?

File: source/non_matching/coupling.cc [hunk 5]
```diff
@@ -285,26 +285,105 @@ namespace NonMatching
...
         // Get a list of outer cells, qpoints and maps.
-        const auto  cpm   = GridTools::compute_point_locations(cache, Xpoints);
-        const auto &cells = std::get<0>(cpm);
-        const auto &qpoints = std::get<1>(cpm);
-        const auto &maps    = std::get<2>(cpm);
-
+        const auto &cells   = cell_container[j];
+        const auto &qpoints = qpoints_container[j];
+        const auto &maps    = maps_container[j];
+        std::cout << "vectors copied" << std::endl;
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: Surely, you wanted to remove these comments, don't you?

File: source/non_matching/coupling.cc [hunk 6]
```diff
@@ -285,26 +285,105 @@ namespace NonMatching
...
+      }
+
     typename DoFHandler<dim1, spacedim>::active_cell_iterator
       cell = immersed_dh.begin_active(),
       endc = immersed_dh.end();
 
-    for (; cell != endc; ++cell)
+    for (unsigned int j = 0; cell != endc; ++cell, ++j)
       {
+        std::cout << j << " out of " << n_active_c << std::endl;
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: Surely, you wanted to remove these comments, don't you?

File: source/non_matching/coupling.cc [hunk 7]
```diff
@@ -285,26 +285,105 @@ namespace NonMatching
...
+            // first we look at the last inserted cell, which is more likely:
+            if (!cell_container[cell_id].empty() &&
+                cell_container[cell_id].back() == all_cells[o])
+              {
+                qpoints_container[cell_id].back().push_back(all_qpoints[o][j]);
+                maps_container[cell_id].back().push_back(n_pt);
+              }
+            else
+              {
+                auto cell_p = std::find(cell_container[cell_id].begin(),
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: This variable can be `const`.

File: source/non_matching/coupling.cc [hunk 8]
```diff
@@ -285,26 +285,105 @@ namespace NonMatching
...
+                                        all_cells[o]);
+                if (cell_p == cell_container[cell_id].end())
+                  {
+                    const std::vector<Point<dim0>> new_qpoint = {
+                      all_qpoints[o][j]};
+                    const std::vector<unsigned int> new_map = {n_pt};
+
+                    cell_container[cell_id].push_back(all_cells[o]);
+                    qpoints_container[cell_id].push_back(new_qpoint);
+                    maps_container[cell_id].push_back(new_map);
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: You could avoid creating the temporary objects if you use `emplace_back` instead.

**Comment by masterleinad** [comment]
> The other reason I post is the following: due to how the function works, I slightly changed the of an addition, thus a test in debug mode fails (differences are in the order 1e-14/-16): would it be ok to make it so the tests prints something like the order of the error?

Yes, that should be fine. If the changes are so small, I wouldn't expect `numdiff` to complain.

File: source/non_matching/coupling.cc (hunk 6 above)
Comment (GivAlz) [reply source/non_matching/coupling.cc]: Oh my, I was sure I had removed them...sorry...cleaning up now!

File: source/non_matching/coupling.cc (hunk 8 above)
Comment (GivAlz) [reply source/non_matching/coupling.cc]: I tried to use emplace_back, but I couldn't understand how to remove completely the temporary vector. This wouldn't compile and I don't understand where I am wrong:
`qpoints_container[cell_id].emplace_back( std::vector<Point<dim0>> ( all_qpoints[o][j]) );`

**Comment by GivAlz** [comment]
Small update, I also did a minor change which avoids std::find to lock at the last element (which I always check by hand, as it is likely to be the correct one).

File: source/non_matching/coupling.cc [hunk 9]
```diff
@@ -330,39 +330,51 @@ namespace NonMatching
...
               {
-                qpoints_container[cell_id].back().push_back(all_qpoints[o][j]);
-                maps_container[cell_id].back().push_back(n_pt);
+                const std::vector<Point<dim0>> new_qpoint = {all_qpoints[o][j]};
+                const std::vector<unsigned int> new_map   = {n_pt};
+                cell_container[cell_id].emplace_back(all_cells[o]);
+                qpoints_container[cell_id].emplace_back(std::begin(new_qpoint),
+                                                        std::end(new_qpoint));
+                maps_container[cell_id].emplace_back(std::begin(new_map),
+                                                     std::end(new_map));
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]:
```
qpoints_container[cell_id].emplace_back(std::vector<Point<dim0>>{all_qpoints[o][j]});
maps_container[cell_id].emplace_back(std::vector<unsigned int>{n_pt});
```
should work

Comment (GivAlz) [reply source/non_matching/coupling.cc]: Thank you :) Now it whould be ok!

File: source/non_matching/coupling.cc [hunk 10]
```diff
@@ -285,25 +285,112 @@ namespace NonMatching
...
+      n_active_c);
+
+    // Cycle over all cells of underling mesh found
+    // call it omesh, elaborating the output
+    for (unsigned int o = 0; o < all_cells.size(); ++o)
+      {
+        for (unsigned int j = 0; j < all_maps[o].size(); ++j)
+          {
+            // Find the index of the "owner" cell and qpoint
+            // w.r.o. the immersed mesh
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: What does "w.r.o" stand for?

File: source/non_matching/coupling.cc [hunk 11]
```diff
@@ -285,25 +285,112 @@ namespace NonMatching
...
+            if (cell_container[cell_id].empty())
+              {
+                cell_container[cell_id].emplace_back(all_cells[o]);
+                qpoints_container[cell_id].emplace_back(
+                  std::vector<Point<dim0>>{all_qpoints[o][j]});
+                maps_container[cell_id].emplace_back(
+                  std::vector<unsigned int>{n_pt});
+              }
+            // If there are already cells, we
+            // begin by looking at the last inserted cell, which is more likely:
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: The formatting of the comment is a bit strange.

File: source/non_matching/coupling.cc [hunk 12]
```diff
@@ -285,25 +285,112 @@ namespace NonMatching
...
+      immersed_dh.get_triangulation().n_active_cells();
+    std::vector<Point<spacedim>> all_points(n_active_c * n_q_points);
+
+    // Collecting all points
+    {
+      unsigned int c = 0;
+      for (const auto &cell : immersed_dh.active_cell_iterators())
+        {
+          // Reinitialize the cell and the fe_values
+          fe_v.reinit(cell);
```
[the remark this answers is not in the document]
Comment (GivAlz) [reply source/non_matching/coupling.cc]: In the sparsity we only need cells, but this time we use fe_v also to compute the matrix entries (line 387 and 438), so it needs to stay out of the `{...}` region.

**Comment by GivAlz** [comment]
Force pushed with some improvements on the comments, code is the same :)

**Comment by masterleinad** [comment]
You probably wanted to squash the last commit or give it a more descriptive name. :wink:

File: source/non_matching/coupling.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply source/non_matching/coupling.cc]: Ah, I see. Thanks!

**Pull request merged**